Add support for --koji simple mode

kontura opened the pull request:
I have tested this on http://download.opensuse.org/distribution/openSUSE-stable/repo/oss/ against koji's mergerepos and there is a difference in inclusion of several packages. The problem are packages which have identical pkgId (hash of the package) as some other package. 

For example there are two identical package files named:
noarch/yast2-trans-de-84.87.20180514.157a0650d-lp150.1.1.noarch.rpm
boot/x86_64/yast2-trans-de.rpm

Basically for this specific repository all the packages in boot/ directory have duplicates.

Is this something we also need to support for issue #137?
I am asking because this is not supported even in the regular --koji mode as they share that part of the code.

tkopecek commented:
Let's not support it for now. Keep 1:1 with koji's mergerepo behaviour.

tkopecek commented:
Sorry, that I've not got to it earlier. I've just confirmed, that behaviour is same with both koji/createrepo_c's mergerepos. Error with same hashes, thus ignoring packages is raised in both (in underlying library), so I'm giving +1 here.

kontura commented:
I have converted long `if () {} else if () {} else if ..` into `switch` statement when selecting merging method, I have also renamed enum values for merge methods to hopefully more reasonable names. 
Newly created header file mergerepo_c.h now contains only necessary definitions and includes.

In addition I have added a new commit, which prevents mergerepo_c from running with undefined command line options.

jrohel commented on src/koji.c:
```diff
@@ -0,0 +1,284 @@
+                // Srpm of the package is not allowed
+                g_debug("Package %s has blocked srpm %s", pkg->name,
+                        pkg->rpm_sourcerpm);
+                return 0;
```
cr_nevra_free(nevra);

jrohel commented on src/mergerepo_c.c:
```diff
@@ -902,10 +544,10 @@ add_package(cr_Package *pkg,
                 // NVR merge method
-                } else if (merge_method == MM_NVR) {
-
+                case MM_NEWEST_FROM_IDENTICAL_NEVRA: {
```
Not identical nevras. Eg. MM_HIGHEST_FROM_NEVRA

kontura commented:
- Renamed misleading `MM_NEWEST_FROM_IDENTICAL_NEVRA` to `MM_WITH_HIGHEST_NEVRA`
- added `cr_nevra_free(nevra)` calls to prevent memory leaks in `koji_allowed`
- removed some trailing spaces
- removed unused variable `blocked_srpms`

Conan-Kudo commented:
@kontura Can you rebase onto current master?

kontura commented:
- added Copyright notices to new files
- moved some more `#includes`  between source and header files

jrohel commented on src/mergerepo_c.c:
```diff
@@ -220,7 +114,10 @@ static GOptionEntry cmd_entries[] =
+       "Enable koji mergerepos behaviour. (Optionally select simple mode with: --simple)", NULL},
+    { "simple", 0, 0, G_OPTION_ARG_NONE, &(_cmd_options.koji_simple),
+       "Enable koji specific simple merge mode where we keep even packages with "
+       "identical NEVRAs. (Doesn't work without --koji)", NULL},
```
There is new text `(Doesn't work without --koji)` added.
There is another option `blocked` with text `Only works with combination with --koji/-k.`
For consistency I prefer the same message for both options.

kontura replied:
Pushed an update with reformatted option text.
